pkg/query-service/app/preferences/model.go

```diff
@@ -0,0 +1,510 @@
+			}
+		}
+	}
+	return nil
+}
+
+func (p *Preference) IsEnabledForScope(scope string) bool {
+	isPreferenceEnabledForGivenScope := false
+	if p.AllowedScopes != nil {
+		for _, allowedScope := range p.AllowedScopes {
+			if allowedScope == scope {
+				isPreferenceEnabledForGivenScope = true
+			}
+		}
+	}
+	return isPreferenceEnabledForGivenScope
+}
+
+func (p *Preference) SanitizeValue(preferenceValue interface{}) interface{} {
+	switch p.ValueType {
+	case PreferenceValueTypeBoolean:
+		if preferenceValue == "1" {
+			return true
+		} else {
+			return false
+		}
+	default:
+		return preferenceValue
+	}
+}
+
+type AllPreferences struct {
+	Preference
+	Value interface{} `json:"value"`
+}
+
+type PreferenceKV struct {
+	PreferenceId    string      `json:"preference_id" db:"preference_id"`
+	PreferenceValue interface{} `json:"preference_value" db:"preference_value"`
+}
+
+type UpdatePreference struct {
+	PreferenceValue interface{} `json:"preference_value"`
+}
+
+var db *sqlx.DB
+
+var preferences []Preference
+
+var preferenceMap map[string]Preference
+
+func InitDB(datasourceName string) error {
+	var err error
+	db, err = sqlx.Open("sqlite3", datasourceName)
+
+	if err != nil {
+		return err
+	}
+
+	// create the user preference table
+	tableSchema := `
+	PRAGMA foreign_keys = ON;
+	CREATE TABLE IF NOT EXISTS user_preference(
+		preference_id TEXT NOT NULL,
+		preference_value TEXT,
+		user_id TEXT NOT NULL,
+		PRIMARY KEY (preference_id,user_id),
+		FOREIGN KEY (user_id)
+			REFERENCES users(id)
+			ON UPDATE CASCADE
+			ON DELETE CASCADE
+	);`
+
+	_, err = db.Exec(tableSchema)
+	if err != nil {
+		return fmt.Errorf("error in creating user_preference table: %s", err.Error())
+	}
+
+	// create the org preference table
+	tableSchema = `
+	PRAGMA foreign_keys = ON;
+	CREATE TABLE IF NOT EXISTS org_preference(
+		preference_id TEXT NOT NULL,
+		preference_value TEXT,
+		org_id TEXT NOT NULL,
+		PRIMARY KEY (preference_id,org_id),
+		FOREIGN KEY (org_id)
+			REFERENCES organizations(id)
+			ON UPDATE CASCADE
+			ON DELETE CASCADE
+	);`
+
+	_, err = db.Exec(tableSchema)
+	if err != nil {
+		return fmt.Errorf("error in creating org_preference table: %s", err.Error())
+	}
+
+	preferenceFromFile, fileErr := fs.ReadFile(os.DirFS("../../pkg/query-service/app/preferences"), "preference.json")
+
+	if fileErr != nil {
+		return fmt.Errorf("error in reading preferences from file : %s", fileErr.Error())
+	}
+
+	if unmarshalErr := json.Unmarshal(preferenceFromFile, &preferences); unmarshalErr != nil {
+		return fmt.Errorf("error in unmarshalling preferences from file : %s", unmarshalErr.Error())
+	}
+
+	preferenceMap = map[string]Preference{}
+	for _, preference := range preferences {
+		_, seen := preferenceMap[preference.Key]
+		if seen {
+			return fmt.Errorf("duplicate preference key in the preferences: %s", preference.Key)
+		}
+		preferenceMap[preference.Key] = preference
+	}
```

Comment:
What's the thought process here? Are we loading some default preferences?

Reply:
If yes, the go binary will ultimately not have the json file and this will throw an error.
You can create a file `map.go` in preferences package and rewrite the json as a `map[string]interface{}`

Reply:
we are not reading the default preferences here but the overall preference entity that is defined across tenants. let me check on loading the JSON part.

Reply:
if i use the `go: embed` directive for this I should be able to load the file right ?

Reply:
You should be able to but the question is why would you want to read a json file, unmarshal it into a struct, then iterate and put into your map when you call initialize the map in the first place?

Reply:
done

Reply:
@grandwizard28 the benefits we get out of using JSON is we can make changes to the config file independent of the go compilation and hence whenever we change the config file we do not have to rebuild the go binary,

Reply:
go:embed will rebuild it!

Reply:
we will not be using `go:embed` for the same rather pass it as a config so that whenever there is a change of addition or deletion for the preferences we do need to rebuild everything.

Reply:
made the changes please check @grandwizard28

Reply:
This is a lot of added complexity for handling a simple thing like constants. Please put it in the code itself.
There is integrity in the preference struct you create and initialize as a constant.

Otherwise, you need to unmarshal from json, put explicit checks for empty fields, correct fields etc.